### Importing relevant libraries

In [1]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd
%matplotlib inline

### Transforming test dataset

In [2]:
df1 = pd.read_csv('Train data.txt')

In [3]:
#Carefuly chose header names by comparing against train data with notepad++

df2 = pd.read_csv('Test data.txt', header = None, names = ['customerID','gender', 'Senior Citizen', 'tenure', 'InternetService', 'Contract','PaymentMethod','MonthlyCharges', 'TotalCharges'])

In [4]:
df2.head()

customerID  gender  Senior Citizen  \
0  9398-MMQTO    Male               0   
1  7619-PLRLP  Female               0   
2  6457-GIRWB    Male               0   
3  6508-NJYRO    Male               0   
4  1450-SKCVI  Female               0   

                                              tenure InternetService  \
0                                                  3     Fiber optic   
1                                                 53             DSL   
2                                                  1     Fiber optic   
3  15 internet service internet service internet ...        One year   
4                                                 56     Fiber optic   

         Contract              PaymentMethod MonthlyCharges  TotalCharges  
0  Month-to-month    Credit card (automatic)          74.45        221.10  
1        One year  Bank transfer (automatic)           74.1       3833.95  
2  Month-to-month           Electronic check          69.35         69.35  
3    Mailed check                       18.8         294.95           NaN  
4  Month-to-month           Electronic check          73.85       4092.85

Visible skew of data from tenure column

In [5]:
df2.isnull().sum()

customerID           0
gender               0
Senior Citizen       0
tenure               0
InternetService      0
Contract             0
PaymentMethod        0
MonthlyCharges       0
TotalCharges       273
dtype: int64

Skew results in 273 missing values for 'Total charges' column

In [6]:
a= (df2.isnull().sum()['TotalCharges']/df2.shape[0])*100
print('Missing value percentage:', a, '%')

Missing value percentage: 19.941563184806427 %


In [7]:
#dropping rows containing missing values to save time cleaning data
df2.dropna(inplace = True)

In [8]:
#Checking for extaordinary unique values in all columns
for i in df2.columns:
    print(df2[i].unique())

['9398-MMQTO' '7619-PLRLP' '6457-GIRWB' ... '1941-HOSAM' '9110-HSGTV'
 '6171-ZTVYB']
['Male' 'Female']
[0 1]
['3' '53' '1' '56' '5' '48' '58' '10' '71' '65' '28 phone service' '67'
 '35' '72 phone service' '61 phone service' '1 phone service' '70'
 '48 phone service' '68' '47' '32 phone service' '13' '12' '9' '38' '42'
 '24' '27' '49 phone service' '61' '50 phone service' '25' '22' '4' '18'
 '51' '62' '16 phone service' '8' '72' '23' '31 phone service'
 '37 phone service' '30' '20' '36' '50' '57' '41' '3 phone service' '64'
 '26 phone service' '40' '34' '63' '44' '6 phone service' '33' '43' '6'
 '59' '14' '19' '2' '57 phone service' '66 phone service'
 '24 phone service' '31' '54' '10 phone service' '11' '15' '32'
 '67 phone service' '29 phone service' '45' '16' '52' '21'
 '14 phone service' '68 phone service' '29' '69 phone service' '17' '46'
 '7' '66' '23 phone service' '37' '21 phone service' '27 phone service'
 '55' '42 phone service' '33 phone service' '69' '60' '13 phone service'

Tenure columns seems like it needs more cleaning

In [9]:
#Applying a simple split method to extract numerical value
df2['tenure'] = df2['tenure'].apply(lambda a: a.split()[0])

In [10]:
#checking values
df2['tenure'].unique()

array(['3', '53', '1', '56', '5', '48', '58', '10', '71', '65', '28',
       '67', '35', '72', '61', '70', '68', '47', '32', '13', '12', '9',
       '38', '42', '24', '27', '49', '50', '25', '22', '4', '18', '51',
       '62', '16', '8', '23', '31', '37', '30', '20', '36', '57', '41',
       '64', '26', '40', '34', '63', '44', '6', '33', '43', '59', '14',
       '19', '2', '66', '54', '11', '15', '29', '45', '52', '21', '69',
       '17', '46', '7', '55', '60', '39'], dtype=object)

In [11]:
df2.head()

customerID  gender  Senior Citizen tenure InternetService        Contract  \
0  9398-MMQTO    Male               0      3     Fiber optic  Month-to-month   
1  7619-PLRLP  Female               0     53             DSL        One year   
2  6457-GIRWB    Male               0      1     Fiber optic  Month-to-month   
4  1450-SKCVI  Female               0     56     Fiber optic  Month-to-month   
5  4710-NKCAW    Male               0      5             DSL  Month-to-month   

               PaymentMethod MonthlyCharges  TotalCharges  
0    Credit card (automatic)          74.45        221.10  
1  Bank transfer (automatic)           74.1       3833.95  
2           Electronic check          69.35         69.35  
4           Electronic check          73.85       4092.85  
5    Credit card (automatic)           64.4        316.90

In [12]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1096 entries, 0 to 1367
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   customerID       1096 non-null   object 
 1   gender           1096 non-null   object 
 2   Senior Citizen   1096 non-null   int64  
 3   tenure           1096 non-null   object 
 4   InternetService  1096 non-null   object 
 5   Contract         1096 non-null   object 
 6   PaymentMethod    1096 non-null   object 
 7   MonthlyCharges   1096 non-null   object 
 8   TotalCharges     1096 non-null   float64
dtypes: float64(1), int64(1), object(7)
memory usage: 85.6+ KB


Tenure and monthly charges have to be changed to int64 and float format respectively

In [13]:
df2['tenure'] = df2['tenure'].astype(int)

In [14]:
df2['MonthlyCharges'] = df2['MonthlyCharges'].astype(float)

In [15]:
#Just Checking
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1096 entries, 0 to 1367
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   customerID       1096 non-null   object 
 1   gender           1096 non-null   object 
 2   Senior Citizen   1096 non-null   int64  
 3   tenure           1096 non-null   int32  
 4   InternetService  1096 non-null   object 
 5   Contract         1096 non-null   object 
 6   PaymentMethod    1096 non-null   object 
 7   MonthlyCharges   1096 non-null   float64
 8   TotalCharges     1096 non-null   float64
dtypes: float64(2), int32(1), int64(1), object(5)
memory usage: 81.3+ KB


### Mirroring manipulation to test data on training data

In [17]:
df1.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [21]:
df1.drop(['Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection','TechSupport', 'StreamingTV', 'StreamingMovies', 'PaperlessBilling'], axis = 1, inplace = True)

In [22]:
df1.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'tenure', 'InternetService',
       'Contract', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [27]:
df1['Churn'].unique()

array(['No', 'Yes'], dtype=object)

In [18]:
df2.columns

Index(['customerID', 'gender', 'Senior Citizen', 'tenure', 'InternetService',
       'Contract', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges'],
      dtype='object')

In [24]:
for i in df1.columns:
    for x in df2.columns:
        if i == x:
            print('True')

True
True
True
True
True
True
True
True


In [26]:
df1.head(20)

customerID  gender  SeniorCitizen  tenure InternetService        Contract  \
0   7590-VHVEG  Female              0       1             DSL  Month-to-month   
1   5575-GNVDE    Male              0      34             DSL        One year   
2   3668-QPYBK    Male              0       2             DSL  Month-to-month   
3   7795-CFOCW    Male              0      45             DSL        One year   
4   9237-HQITU  Female              0       2     Fiber optic  Month-to-month   
5   9305-CDSKC  Female              0       8     Fiber optic  Month-to-month   
6   1452-KIOVK    Male              0      22     Fiber optic  Month-to-month   
7   6713-OKOMC  Female              0      10             DSL  Month-to-month   
8   7892-POOKP  Female              0      28     Fiber optic  Month-to-month   
9   6388-TABGU    Male              0      62             DSL        One year   
10  9763-GRSKD    Male              0      13             DSL  Month-to-month   
11  7469-LKBCI    Male              0      16              No        Two year   
12  8091-TTVAX    Male              0      58     Fiber optic        One year   
13  0280-XJGEX    Male              0      49     Fiber optic  Month-to-month   
14  5129-JLPIS    Male              0      25     Fiber optic  Month-to-month   
15  3655-SNQYZ  Female              0      69     Fiber optic        Two year   
16  8191-XWSZG  Female              0      52              No        One year   
17  9959-WOFKT    Male              0      71     Fiber optic        Two year   
18  4190-MFLUW  Female              0      10             DSL  Month-to-month   
19  4183-MYFRB  Female              0      21     Fiber optic  Month-to-month   

                PaymentMethod  MonthlyCharges TotalCharges Churn  
0            Electronic check           29.85        29.85    No  
1                Mailed check           56.95       1889.5    No  
2                Mailed check           53.85       108.15   Yes  
3   Bank transfer (automatic)           42.30      1840.75    No  
4            Electronic check           70.70       151.65   Yes  
5            Electronic check           99.65        820.5   Yes  
6     Credit card (automatic)           89.10       1949.4    No  
7                Mailed check           29.75        301.9    No  
8            Electronic check          104.80      3046.05   Yes  
9   Bank transfer (automatic)           56.15      3487.95    No  
10               Mailed check           49.95       587.45    No  
11    Credit card (automatic)           18.95        326.8    No  
12    Credit card (automatic)          100.35       5681.1    No  
13  Bank transfer (automatic)          103.70       5036.3   Yes  
14           Electronic check          105.50      2686.05    No  
15    Credit card (automatic)          113.25      7895.15    No  
16               Mailed check           20.65      1022.95    No  
17  Bank transfer (automatic)          106.70      7382.25    No  
18    Credit card (automatic)           55.20       528.35   Yes  
19           Electronic check           90.05       1862.9    No